facilitator/src/task.rs

```diff
@@ -1,22 +1,39 @@
 mod pubsub;
 mod sqs;
 
-use anyhow::{Context, Result};
+use anyhow::Result;
 use serde::Deserialize;
 use std::{
     fmt,
     fmt::{Debug, Display},
 };
 
-pub use pubsub::PubSubTaskQueue;
-pub use sqs::SqsTaskQueue;
+pub use pubsub::GcpPubSubTaskQueue;
+pub use sqs::AwsSqsTaskQueue;
 
-/// Represents a task that can be assigned to a worker
-pub trait Task: Debug + Display + Sized {
-    /// Attempts to deserialize a task structure from the provided JSON string
-    fn from_json(json: &[u8]) -> Result<Self>;
+/// A queue of tasks to be executed
+pub trait TaskQueue<T: Task>: Debug {
+    /// Get a task to execute. If a task to run is found, returns Ok(Some(T)).
+    /// If a task is successfully checked for but there is no work available,
+    /// returns Ok(None). Returns Err(e) if something goes wrong.
+    /// Once the task has been successfully completed, the TaskHandle should be
+    /// passed to acknowledge_task to permanently remove the task. If
+    /// acknowledge_task is never called, the task will eventually be
+    /// re-delivered via dequeue().
+    fn dequeue(&mut self) -> Result<Option<TaskHandle<T>>>;
+
+    /// Signal to the task queue that the task has been handled and should be
+    /// removed from the queue.
+    fn acknowledge_task(&mut self, handle: TaskHandle<T>) -> Result<()>;
+
+    /// Signal to the task queue that the task was not handled and should be
+    /// retried later.
+    fn nacknowledge_task(&mut self, handle: TaskHandle<T>) -> Result<()>;
```

Comment:
Love this name :-D

Reply:
It's easy to mix up though, ack vs nack is easy to distinguish, acknowledge vs nacknowledge less so.